Userland/Libraries/LibWeb/MimeSniff/Resource.cpp

```diff
@@ -0,0 +1,625 @@
+/*
+ * Copyright (c) 2023, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "Resource.h"
+
+namespace {
+
+using namespace Web::MimeSniff;
+using Byte = u8;
+
+class BytePatternTableRow {
+public:
+    BytePatternTableRow(ReadonlyBytes byte_pattern, ReadonlyBytes pattern_mask, HashTable<Byte> const& ignored_leading_bytes, StringView mime_type, bool is_tag_terminated = false)
+        : m_byte_pattern(byte_pattern)
+        , m_pattern_mask(pattern_mask)
+        , m_ignored_leading_bytes(ignored_leading_bytes)
+        , m_mime_type(mime_type)
+        , m_is_tag_terminated(is_tag_terminated)
+    {
+    }
+
+    ~BytePatternTableRow() = default;
+
+    ReadonlyBytes byte_pattern() const { return m_byte_pattern; }
+    ReadonlyBytes pattern_mask() const { return m_pattern_mask; }
+    HashTable<Byte> const& ignored_leading_bytes() const { return m_ignored_leading_bytes; }
+    StringView mime_type() const { return m_mime_type; }
+    bool is_tag_terminated() const { return m_is_tag_terminated; }
+
+private:
+    ReadonlyBytes m_byte_pattern;
+    ReadonlyBytes m_pattern_mask;
+    HashTable<Byte> const& m_ignored_leading_bytes;
```

Comment:
these leading bytes are very short arrays, so IMO it would be fine to use plain arrays here, as these might actually be faster. but someone more knowledgeable should chime in.

Reply:
Yes, this is only ever really used with the matching an [rules for identifying an unknown MIME type algorithm](https://mimesniff.spec.whatwg.org/#identifying-a-resource-with-an-unknown-mime-type) when we need to ignore their definition of whitespace bytes and I do agree with you that this may be overkill with such a small input.

The reason I made this a `HashTable<>` is the [pattern matching algo](https://mimesniff.spec.whatwg.org/#pattern-matching-algorithm) has its `ignored` parameter as an ordered set (not sure why ordered because order doesn't matter here) and I wanted to abide by it. I'll change it to an `Array<>`

Reply:
Used a `Vector<>` instead because of the pattern matching algo